### PR TITLE
add nproc shim for macOS

### PR DESCRIPTION
**Pull request by jedahan**
This seems pretty hacky but was needed for me to run `./build/dev.sh` on my mac. Open to other suggestions.

### CODE REVIEW

**Comment by andychu** [comment]
Thanks for the patch!  I think there are going to be many other issues, which I hint at in the first sentence here:

https://github.com/oilshell/oil/wiki/Contributing

Maybe I should make it more clear, but Oil development is very test driven, and the testing involves running other shells.  bash and zsh run on OS X just fine I'm sure, although we need the specific versions in `test/spec-bin.sh` for the test to be green.  And I'm not sure of mksh and dash work on OS X.

Long story short, nobody is developing on Mac now and I think it will involve a lot of yak shaving :-/  It's a bit unfortunate but the reasons are somewhat fundamental I think.

I use VirtualBox running Ubuntu when I'm on a Mac (which I was for the whole summer at RC!)

Someone suggested building/running in a Docker container on OS X.  That might be feasible, but I haven't looked into it personally.

**Comment by jedahan** [comment]
Welp I'll journal here what yak-shaving I get done and we can see if it seems like progress, or at least help future folks playing around with it on macOS. If it ends up being more of a maintainer burden/distraction feel free to close.

* `bin/osh` runs after `build/dev.sh minimal`
* had to install gawk with `brew install gawk` to get tests to run
* 2 failures in `test/unit.sh all`
* 12 failures in `test/spec.sh all` (6 osh 6 bash, all match)
* installed mksh with `brew install mksh` and reran smoke test
* installed dash with `brew install dash` and reran smoke test
```zsh
△ test/spec.sh smoke                                             ~/code/oil nproc-shim 8h ⬢  ⚙
smoke.test.sh
case    line    dash    bash    mksh    osh
  0       5     pass    pass    pass    pass    builtin
  1       9     pass    pass    pass    pass    command sub
  2      13     pass    pass    pass    pass    command sub with builtin
  3      17     FAIL    FAIL    FAIL    FAIL    pipeline
  4      21     FAIL    FAIL    FAIL    FAIL    pipeline with builtin
  5      25     pass    pass    pass    pass    and-or chains
  6      42     FAIL    FAIL    FAIL    FAIL    here doc with var
  7      50     FAIL    FAIL    FAIL    FAIL    here doc without var
  8      57     FAIL    pass    pass    pass    here doc with builtin
  9      64     pass    pass    pass    pass    Redirect external command
 10      70     pass    pass    pass    pass    Redirect with builtin
 11      75     FAIL    FAIL    FAIL    FAIL    Here doc with redirect
 12      83     pass    pass    pass    pass    "$@" "$*"
 13      90     pass    pass    pass    pass    $@ $*
 14      97     FAIL    FAIL    FAIL    FAIL    failed command
 15     101     pass    pass    pass    pass    subshell
 16     109     pass    pass    pass    pass    for loop
 17     121     pass    pass    pass    pass    vars
47 passed, 0 ok, 0 known unimplemented, 0 known bugs, 25 failed, 0 skipped

FATAL: 25 tests failed (6 osh failures)
```
bash 4.4.23
zsh 5.6.2
dash 0.5.10.2
mksh 56c

now trying out `build/test-bin.sh build-all` to see if there are any discrepancies, and learn what might work/not work for getting tests to go green.

**Comment by jedahan** [comment]
woah, funny thing, I got almost all tests to pass by switching to coreutils with `brew install coreutils` and `export PATH="/usr/local/opt/coreutils/libexec/gnubin:$PATH"`:
```shell
▲ test/spec.sh smoke                                             ~/code/oil nproc-shim 8h ⬢  ⚙
/usr/local/opt/coreutils/libexec/gnubin/nproc
smoke.test.sh
case    line    dash    bash    mksh    osh
  0       5     pass    pass    pass    pass    builtin
  1       9     pass    pass    pass    pass    command sub
  2      13     pass    pass    pass    pass    command sub with builtin
  3      17     pass    pass    pass    pass    pipeline
  4      21     pass    pass    pass    pass    pipeline with builtin
  5      25     pass    pass    pass    pass    and-or chains
  6      42     pass    pass    pass    pass    here doc with var
  7      50     pass    pass    pass    pass    here doc without var
  8      57     FAIL    pass    pass    pass    here doc with builtin
  9      64     pass    pass    pass    pass    Redirect external command
 10      70     pass    pass    pass    pass    Redirect with builtin
 11      75     pass    pass    pass    pass    Here doc with redirect
 12      83     pass    pass    pass    pass    "$@" "$*"
 13      90     pass    pass    pass    pass    $@ $*
 14      97     pass    pass    pass    pass    failed command
 15     101     pass    pass    pass    pass    subshell
 16     109     pass    pass    pass    pass    for loop
 17     121     pass    pass    pass    pass    vars
71 passed, 0 ok, 0 known unimplemented, 0 known bugs, 1 failed, 0 skipped

FATAL: 1 tests failed (0 osh failures)
```

**Comment by jedahan** [comment]
Seems like dash 0.5.10.2 is double escaping the `=`:
```
△ test/spec.sh smoke --range 8 --verbose                         ~/code/oil nproc-shim 8h ⬢  ⚙
/usr/local/opt/coreutils/libexec/gnubin/nproc
smoke.test.sh
case    line    dash    bash    mksh    osh
  8      57     FAIL    pass    pass    pass    here doc with builtin
case: 8
[dash stdout] Expected 'var = value\n', got 'var \\= value\n'
dash stdout:
var \= value

dash stderr:


3 passed, 0 ok, 0 known unimplemented, 0 known bugs, 1 failed, 17 skipped

FATAL: 1 tests failed (0 osh failures)
```

**Comment by andychu** [comment]
OK interesting!  `coreutils` makes sense because I remember there are some significant differences with the BSD-derived utils on OS X.

Well enough people have tried to develop on Mac, so if you want to dump a shell script like `build/dev-osx.sh` or something with some brew commands that seems useful.  If nothing else to document it for the next person :)

`build/dev.sh` has an `ubuntu-deps` function, so there could the analogous thing for OS X and brew.

I am not sure there won't be other hurdles but I think that's good progress.  If a couple tests are broken on OS X due to version that may not hinder people from developing.

File: build/dev.sh [hunk 1]
```diff
@@ -33,6 +33,26 @@ else
...
+  command -v gcat 2> /dev/null || brew install coreutils
+  command -v python2 2> /dev/null || brew install python@2
+  cat --version 2>/dev/null | grep -q GNU || {
+    echo 'exporting path to point to gnu coreutils for tests'
+    echo 'you may want to add this your bashrc or zshrc:'
+    echo 'export PATH="/usr/local/opt/coreutils/libexec/gnubin:$PATH"'
+    export PATH='/usr/local/opt/coreutils/libexec/gnubin:$PATH'
+  }
+
+  test/spec.sh install-shells
```
Comment (andychu) [inline build/dev.sh]: This should probably be removed as it is Linux-specific.
